automata/fa/dfa.py

```diff
@@ -710,17 +710,159 @@ def maximum_word_length(self):
             return float('inf')
 
     @classmethod
-    def from_finite_language(cls, language, input_symbols):
+    def contains_substring(cls, input_symbols, substring):
+        """
+        Directly computes the minimal DFA recognizing strings containing the
+        given substring.
+        """
+        prefixes = [substring[:i] for i in range(len(substring))]
+
+        transitions = {prefix: dict() for prefix in prefixes}
+        transitions[substring] = {
+            symbol: substring for symbol in input_symbols
+        }
+
+        for prefix in prefixes:
+            prefix_dict = transitions.setdefault(prefix, dict())
+
+            for symbol in input_symbols:
+                # Look for next state after reading in the given input symbol
+                possible_suffix = prefix + symbol
+
+                # This while loop will always terminate, since the empty string is in the dict
+                while possible_suffix not in transitions:
+                    possible_suffix = possible_suffix[1:]
+
+                prefix_dict[symbol] = possible_suffix
+
+        return cls(
+            states=set(transitions.keys()),
+            input_symbols=input_symbols,
+            transitions=transitions,
+            initial_state='',
+            final_states={substring},
+        )
+
+    @classmethod
+    def contains_subsequence(cls, input_symbols, subsequence):
+        """
+        Creates a DFA which accepts all words which contain a specific subsequence of symbols
+        """
+        transitions = {0: {symbol: 0 for symbol in input_symbols}}
+
+        for prev_state, symbol in enumerate(subsequence):
+            next_state = prev_state + 1
+            transitions[next_state] = {symbol: next_state for symbol in input_symbols}
+            transitions[prev_state][symbol] = next_state
+
+        return cls(
+            states=set(transitions.keys()),
+            input_symbols=input_symbols,
+            transitions=transitions,
+            initial_state=0,
+            final_states={len(subsequence)},
+        )
+
+    @classmethod
+    def of_length(cls, input_symbols, min_length=0, max_length=float('inf')):
```

Comment:
Wondering if keyword-only args would be good to use here, at least for `min_length` and `max_length. I think it would make your README example a bit cleaner:

```python
def of_length(cls, input_symbols, *, min_length=0, max_length=float('inf')):
    ...
```

```python
dfa = DFA.of_length({'0', '1'}, min_length=4, max_length=float('inf'))
```

@eliotwrobson, what do you think?

Reply:
Sounds good, in keeping with the conventions of other functions. I think this got mentioned before, but I think that `None` is a better default sentinel value than `float('inf')`.

Reply:
We should be able to pass whatever is returned by the word length functions in to this function, in my opinion. The default sentinel value can be None but it must also support float('inf') if the `maximum_word_length` function returns float('inf').
If we switch to None then we still are passing in two different types.

Reply:
Updated to keyword arguments

Reply:
Right, but I think `None` is known to be a sentinel value (enough that Python has native support for it with the `Optional` type), while the `float('inf')` is requiring the user to passing _one specific value_ of the `float` type (they can't put in `3.5`, for example). I think it's actually worth considering whether we should remove all of the `float('inf')` return types altogether for this. Ultimately it's up to @caleb531, since is more about how the library is organized rather than any serious algorithmic considerations.

Reply:
I would agree with @eliotwrobson that `None` is preferable. It is a Python convention, and semantically, it implies that "there is no maximum length" rather than "the maximum length is infinity" (the former of which makes more sense to me).

Plus, it is unnatural to me to represent a length value with a float, per @eliotwrobson's other point. Sure, we could add a library to represent infinity as an integer, but that seems a bit overkill to me, and erodes my first argument above.

Reply:
I'm fine with using `None`, there's just pros and cons with each approach.
`None` also semantically makes sense for the empty language, but there it means something completely different.
I think leaving that case as an exception though makes more sense since it truly is an exception.

I don't really like using floating points at all when I can avoid it. The reason I chose it originally was simplicity, you can compare using the `<` operator and everything will work the same without special casing.

I think we should also avoid adding dependencies unless it is heavily relied upon or adds great value. If we wanted to go that way, creating an integer data type with a sentinel infinity value is relatively simple and does not warrant adding a dependency in my opinion.

I will change all the logic to use `None` then.

Reply:
Now my only concern is the `cardinality` function. I feel like `None` is not the proper value to return so I'm inclined to have it raise a `ValueError`, just as `len` must do.

Reply:
Updated